xt::reducer doesn't work since v0.12

I followed the example, the code works in v0.11, but doesn't work in any version after, including master.

```
    xt::xarray<double> arr = xt::random::rand<double>({3, 2, 4, 6, 5});
    xt::xarray<double> res = xt::reduce([](double a, double b) { return a*a + b*b; },
                                        arr,
                                        {1, 3});
```

errors like

```
xtensor/xreducer.hpp: In instantiation of ‘class xt::xreducer<main(int, char**)::<lambda(double, double)>, const xt::xarray_container<xt::uvector<double, std::allocator<double> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&, std::array<long unsigned int, 2> >’:
xtensor/xreducer.hpp:530:20:   required from ‘auto xt::detail::reduce_impl(F&&, E&&, X&&, xt::evaluation_strategy::lazy) [with F = main(int, char**)::<lambda(double, double)>; E = xt::xarray_container<xt::uvector<double, std::allocator<double> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&; X = std::array<long unsigned int, 2>]’
xtensor/xreducer.hpp:589:35:   required from ‘auto xt::reduce(F&&, E&&, const I (&)[N], EVS) [with F = main(int, char**)::<lambda(double, double)>; E = xt::xarray_container<xt::uvector<double, std::allocator<double> >, (xt::layout_type)1, xt::svector<long unsigned int, 4, std::allocator<long unsigned int>, true>, xt::xtensor_expression_tag>&; I = int; long unsigned int N = 2; EVS = xt::evaluation_strategy::lazy]’
/home/colinf/Dropbox/Gambit/Gitlab/xtensor_test/src/test.cpp:62:47:   required from here
xtensor/xreducer.hpp:431:82: error: no type named ‘reduce_functor_type’ in ‘std::decay_t<main(int, char**)::<lambda(double, double)> > {aka struct main(int, char**)::<lambda(double, double)>}’
         using reduce_functor_type = typename std::decay_t<F>::reduce_functor_type;
                                                                                  ^
xtensor/xreducer.hpp:432:78: error: no type named ‘init_functor_type’ in ‘std::decay_t<main(int, char**)::<lambda(double, double)> > {aka struct main(int, char**)::<lambda(double, double)>}’
         using init_functor_type = typename std::decay_t<F>::init_functor_type;
                                                                              ^
xtensor/xreducer.hpp:433:80: error: no type named ‘merge_functor_type’ in ‘std::decay_t<main(int, char**)::<lambda(double, double)> > {aka struct main(int, char**)::<lambda(double, double)>}’
         using merge_functor_type = typename std::decay_t<F>::merge_functor_type;
                                                                                ^
xtensor/xreducer.hpp:441:50: error: no type named ‘reduce_functor_type’ in ‘std::decay<main(int, char**)::<lambda(double, double)> >::type {aka struct main(int, char**)::<lambda(double, double)>}’
```

--- a/docs/source/operator.rst
+++ b/docs/source/operator.rst
@@ -191,6 +191,22 @@ You can also call the ``reduce`` generator with your own reducing function:
                                         arr,
                                         {1, 3});
 
+The reduce generator also accepts a ``xreducer_functors`` object, a tuple of three functions
+(one for reducing, one for initialization and one for merging). A generator is provided to
+build the ``xreducer_functors`` object, the last function can be omitted:
+
+.. code::
+
+    #include "xtensor/xarray.hpp"
+    #include "xtensor/xreducer.hpp"
+
+    xt::xarray<double> arr = some_init_function({3, 2, 4, 6, 5});
+    xt::xarray<double> res = xt::reduce(xt::make_xreducer_functor([](double a, double b) { return a*a + b*b; },
+                                                                  [](double a) { return a * 2; })
+                                        arr,
+                                        {1, 3});
+
+
 Accumulators
 ------------
 

--- a/include/xtensor/xreducer.hpp
+++ b/include/xtensor/xreducer.hpp
@@ -541,6 +541,23 @@ namespace xt
 
 #define DEFAULT_STRATEGY_REDUCERS evaluation_strategy::lazy
 
+    namespace detail
+    {
+        template <class T>
+        struct is_xreducer_functors_impl : std::false_type
+        {
+        };
+
+        template <class RF, class IF, class MF>
+        struct is_xreducer_functors_impl<xreducer_functors<RF, IF, MF>>
+            : std::true_type
+        {
+        };
+
+        template <class T>
+        using is_xreducer_functors = is_xreducer_functors_impl<std::decay_t<T>>;
+    }
+
     /**
      * @brief Returns an \ref xexpression applying the specified reducing
      * function to an expression over the given axes.
@@ -555,14 +572,24 @@ namespace xt
      */
 
     template <class F, class E, class X, class EVS = DEFAULT_STRATEGY_REDUCERS,
-              XTL_REQUIRES(xtl::negation<is_evaluation_strategy<X>>)>
+              XTL_REQUIRES(xtl::negation<is_evaluation_strategy<X>>,
+                           detail::is_xreducer_functors<F>)>
     inline auto reduce(F&& f, E&& e, X&& axes, EVS evaluation_strategy = EVS())
     {
         return detail::reduce_impl(std::forward<F>(f), std::forward<E>(e), std::forward<X>(axes), evaluation_strategy);
     }
 
+    template <class F, class E, class X, class EVS = DEFAULT_STRATEGY_REDUCERS,
+              XTL_REQUIRES(xtl::negation<is_evaluation_strategy<X>>,
+                           xtl::negation<detail::is_xreducer_functors<F>>)>
+    inline auto reduce(F&& f, E&& e, X&& axes, EVS evaluation_strategy = EVS())
+    {
+        return reduce(make_xreducer_functor(std::forward<F>(f)), std::forward<E>(e), std::forward<X>(axes), evaluation_strategy);
+    }
+
     template <class F, class E, class EVS = DEFAULT_STRATEGY_REDUCERS,
-              XTL_REQUIRES(is_evaluation_strategy<EVS>)>
+              XTL_REQUIRES(is_evaluation_strategy<EVS>,
+                           detail::is_xreducer_functors<F>)>
     inline auto reduce(F&& f, E&& e, EVS evaluation_strategy = EVS())
     {
         xindex_type_t<typename std::decay_t<E>::shape_type> ar;
@@ -571,8 +598,17 @@ namespace xt
         return detail::reduce_impl(std::forward<F>(f), std::forward<E>(e), std::move(ar), evaluation_strategy);
     }
 
+    template <class F, class E, class EVS = DEFAULT_STRATEGY_REDUCERS,
+              XTL_REQUIRES(is_evaluation_strategy<EVS>,
+                           xtl::negation<detail::is_xreducer_functors<F>>)>
+    inline auto reduce(F&& f, E&& e, EVS evaluation_strategy = EVS())
+    {
+        return reduce(make_xreducer_functor(std::forward<F>(f)), std::forward<E>(e), evaluation_strategy);
+    }
+
 #ifdef X_OLD_CLANG
-    template <class F, class E, class I, class EVS = DEFAULT_STRATEGY_REDUCERS>
+    template <class F, class E, class I, class EVS = DEFAULT_STRATEGY_REDUCERS,
+              XTL_REQUIRES(detail::is_xreducer_functors<F>)>
     inline auto reduce(F&& f, E&& e, std::initializer_list<I> axes, EVS evaluation_strategy = EVS())
     {
         using axes_type = std::vector<std::size_t>;
@@ -580,13 +616,26 @@ namespace xt
         using reducer_type = xreducer<F, const_xclosure_t<E>, axes_type>;
         return detail::reduce_impl(std::forward<F>(f), std::forward<E>(e), std::move(ax), evaluation_strategy);
     }
+    template <class F, class E, class I, class EVS = DEFAULT_STRATEGY_REDUCERS,
+              XTL_REQUIRES(xtl::negation<detail::is_xreducer_functors<F>>)>
+    inline auto reduce(F&& f, E&& e, std::initializer_list<I> axes, EVS evaluation_strategy = EVS())
+    {
+        return reduce(make_xreducer_functor(std::forward<F>(f)), std::forward<E>(e), axes, evaluation_strategy);
+    }
 #else
-    template <class F, class E, class I, std::size_t N, class EVS = DEFAULT_STRATEGY_REDUCERS>
+    template <class F, class E, class I, std::size_t N, class EVS = DEFAULT_STRATEGY_REDUCERS,
+              XTL_REQUIRES(detail::is_xreducer_functors<F>)>
     inline auto reduce(F&& f, E&& e, const I (&axes)[N], EVS evaluation_strategy = EVS())
     {
         using axes_type = std::array<std::size_t, N>;
         auto ax = xt::forward_normalize<axes_type>(e, axes);
         return detail::reduce_impl(std::forward<F>(f), std::forward<E>(e), std::move(ax), evaluation_strategy);
+    }
+    template <class F, class E, class I, std::size_t N, class EVS = DEFAULT_STRATEGY_REDUCERS,
+              XTL_REQUIRES(xtl::negation<detail::is_xreducer_functors<F>>)>
+    inline auto reduce(F&& f, E&& e, const I (&axes)[N], EVS evaluation_strategy = EVS())
+    {
+        return reduce(make_xreducer_functor(std::forward<F>(f)), std::forward<E>(e), axes, evaluation_strategy);
     }
 #endif
 

--- a/test/test_xreducer.cpp
+++ b/test/test_xreducer.cpp
@@ -56,6 +56,10 @@ namespace xt
         xt::xarray<int> expect = {6, 15};
         EXPECT_EQ(a_sums, expect);
         EXPECT_EQ(a_sums2, expect);
+
+        xt::xarray<int> a_sums3 = xt::reduce(sum, a, {1});
+        EXPECT_EQ(a_sums3, expect);
+
     }
 
     TEST(xreducer, errors)

